cpp/src/indexing/point_quadtree.cu

```diff
@@ -0,0 +1,957 @@
+    // initial point keys (sorted z-order Morton codes): [1, 3, 4, 7, 8] |
+    //                                                    ▾  ▾  ▾  ▾  ▾  |
+    //              .-------------------------------------`  :  :  :  :  |
+    //              :          .-----------------------------`  :  :  :  |
+    //              :          :          .---------------------`  :  :  |
+    //              :          :          :          .-------------`  :  |
+    //              :          :          :          :          .-----`  |
+    // ----------|--+----------+----------+----------+----------+--------|
+    //     level |        quadtree nodes [key, num children, num points] |
+    // ----------|--+----------+----------+----------+----------+--------|
+    //           |  ▾          ▾          ▾          ▾          ▾        |
+    // (start) 3 | [1, 0, 1]  [3, 0, 1]  [4, 0, 1]  [7, 0, 1]  [8, 0, 1] |
+    //           |     ▾          :          ▾          :          :     |
+    //           |     `---------.▾          `---------.▾          ▾     |
+    //         2 |            [0, 2, 2]             [1, 2, 2]  [2, 1, 1] |
+    //           |                ▾                     :          :     |
+    //           |                `--------------------.▾          ▾     |
+    //         1 |                                  [0, 2, 4]  [1, 1, 1] |
+    //           |                                      ▾          :     |
+    //           |                                      `---------.▾     |
+    //   (end) 0 |                                             [0, 2, 5] |
+    //                                                           (root)  |
+    //
+    std::tie(num_parent_nodes, quad_tree_size, begin_pos, end_pos) =
+      construct_quadrants(num_levels,
+                          num_top_quads,
+                          quad_keys.begin(),
+                          quad_point_count.begin(),
+                          quad_child_count.begin(),
+                          stream);
+
+    // Shrink vectors' underlying device allocations to reduce peak memory usage
+    shrink_vector(quad_keys, quad_tree_size);
+    shrink_vector(quad_point_count, quad_tree_size);
+    shrink_vector(quad_child_count, quad_tree_size);
+
+    // Optimization: return early if the top level nodes are all leaves
+    if (num_parent_nodes <= 0) {
+      return std::make_tuple(quad_keys,
+                             quad_point_count,
+                             quad_child_count,
+                             quad_levels,
+                             num_top_quads,
+                             num_parent_nodes,
+                             0);
+    }
+
+    rmm::device_vector<uint32_t> quad_keys_reversed{};
+    rmm::device_vector<uint32_t> quad_point_count_reversed{};
+    rmm::device_vector<uint32_t> quad_child_count_reversed{};
+    rmm::device_vector<int8_t> quad_levels{};
+
+    // Reverse the quadtree nodes for easier manipulation (skips the root node)
```

Comment:
Explain why reversing them makes manipulation easier. What type of manipulation?